src/stream/mod.rs

```diff
@@ -243,16 +247,90 @@ impl<R: Read + Seek> Reader<R> {
     }
 
     fn is_binary(reader: &mut R) -> Result<bool, Error> {
-        fn from_io_offset_0(err: io::Error) -> Error {
-            ErrorKind::Io(err).with_byte_offset(0)
+        Self::rewind(reader)?;
+        let is_binary = Self::reader_matches(reader, b"bplist00")?;
+        Self::rewind(reader)?;
+        Ok(is_binary)
+    }
+
+    fn is_xml(reader: &mut R) -> Result<bool, Error> {
+        const UTF32_BE_BOM: [u8; 4] = [0, 0, 0xfe, 0xff];
+        const UTF32_LE_BOM: [u8; 4] = [0xff, 0xfe, 0, 0];
+        const UTF32_2143_BOM: [u8; 4] = [0, 0, 0xff, 0xfe];
+        const UTF32_3412_BOM: [u8; 4] = [0xfe, 0xff, 0, 0];
+        const UTF8_BOM: [u8; 3] = [0xef, 0xbb, 0xbf];
+        const UTF16_BE_BOM: [u8; 2] = [0xfe, 0xff];
+        const UTF16_LE_BOM: [u8; 2] = [0xff, 0xfe];
+
+        const BOMS: [&[u8]; 7] = [
+            UTF32_BE_BOM.as_slice(),
+            UTF32_LE_BOM.as_slice(),
+            UTF32_2143_BOM.as_slice(),
+            UTF32_3412_BOM.as_slice(),
+            UTF8_BOM.as_slice(),
+            UTF16_BE_BOM.as_slice(),
+            UTF16_LE_BOM.as_slice(),
+        ];
+
+        for bom in BOMS {
+            Self::rewind(reader)?;
+            let matches = Self::reader_matches(reader, bom)?;
+
+            if matches {
+                Self::rewind(reader)?;
+                return Ok(true);
+            }
         }
 
-        reader.seek(SeekFrom::Start(0)).map_err(from_io_offset_0)?;
-        let mut magic = [0; 8];
-        reader.read_exact(&mut magic).map_err(from_io_offset_0)?;
-        reader.seek(SeekFrom::Start(0)).map_err(from_io_offset_0)?;
+        Self::rewind(reader)?;
+
+        let is_xml = loop {
+            let byte = Self::next_byte(reader)?;
+            if byte.is_ascii_whitespace() {
+                continue;
+            }
+
+            if byte == b'<' && Self::reader_matches(reader, b"?xml")? {
```

Comment:
The new `xml_with_bom_and_whitespace` test proves that a BOM and whitespace can precede `<?xml`, and the `dictionary_deserialize_dictionary_in_struct` implicitly proves that whitespace and no BOM can precede `<?xml`, I can make an explicit test for it if you prefer.

All XML documents must now begin with:
1. An optional BOM
2. Optional whitespace
3. `<?xml`

Otherwise they'll be treated as ASCII.